src/elliptic/curves/traits.rs

```diff
@@ -5,48 +5,215 @@
     License MIT: <https://github.com/KZen-networks/curv/blob/master/LICENSE>
 */
 
-use std::ops::{Add, Mul};
+use std::fmt;
+
+use serde::{Deserialize, Serialize};
+use zeroize::Zeroize;
 
 use crate::BigInt;
-use crate::ErrorKey;
 
-pub trait ECScalar: Mul<Output = Self> + Add<Output = Self> + Sized {
-    type SecretKey;
+/// Elliptic curve implementation
+///
+/// Refers to according implementation of [ECPoint] and [ECScalar].
+pub trait Curve {
+    type Point: ECPoint<Scalar = Self::Scalar>;
+    type Scalar: ECScalar;
+
+    /// Returns canonical name for this curve
+    fn curve_name() -> &'static str;
+}
+
+/// Scalar value modulus [curve order](Self::curve_order)
+///
+/// ## Note
+/// This is a low-level trait, you should not use it directly. See wrappers [Point], [PointZ],
+/// [Scalar], [ScalarZ].
+///
+/// [Point]: super::wrappers::Point
+/// [PointZ]: super::wrappers::PointZ
+/// [Scalar]: super::wrappers::Scalar
+/// [ScalarZ]: super::wrappers::ScalarZ
+///
+/// Trait exposes various methods to manipulate scalars. Scalar can be zero. Scalar must zeroize its
+/// value on drop.
+pub trait ECScalar: Clone + PartialEq + fmt::Debug + 'static {
+    /// Underlying scalar type that can be retrieved in case of missing methods in this trait
+    type Underlying;
+
+    /// Samples a random scalar
+    fn random() -> Self;
+
+    /// Constructs a zero scalar
+    fn zero() -> Self;
+    /// Checks if the scalar equals to zero
+    fn is_zero(&self) -> bool;
+
+    /// Constructs a scalar `n % curve_order`
+    fn from_bigint(n: &BigInt) -> Self;
+    /// Converts a scalar to BigInt
+    fn to_bigint(&self) -> BigInt;
+
+    /// Calculates `(self + other) mod curve_order`
+    fn add(&self, other: &Self) -> Self;
+    /// Calculates `(self * other) mod curve_order`
+    fn mul(&self, other: &Self) -> Self;
+    /// Calculates `(self - other) mod curve_order`
+    fn sub(&self, other: &Self) -> Self;
+    /// Calculates `-self mod curve_order`
+    fn neg(&self) -> Self;
```

Comment:
Are you asking if we can add bound `ops::Neg` to ECPoint instead of having `.neg()` method inside ECPoint? I think it might be confusing - other arithmetic operators (add, sub, mul) are defined via trait method, and there's no way to replace them with trait bounds

Reply:
No this was answered when I saw the `matrix` macro, just forgot to delete this comment